postgres/datadog_checks/postgres/statement_samples.py

```diff
@@ -0,0 +1,355 @@
+        )
+        self._check.histogram("dd.postgres.get_new_pg_stat_activity.rows", len(rows), tags=self._tags)
+        self._log.debug("Loaded %s rows from %s", len(rows), self._config.pg_stat_activity_view)
+        return rows
+
+    def _filter_valid_statement_rows(self, rows):
+        insufficient_privilege_count = 0
+        total_count = 0
+        for row in rows:
+            total_count += 1
+            if not row['datname']:
+                continue
+            query = row['query']
+            if not query:
+                continue
+            if query == '<insufficient privilege>':
+                insufficient_privilege_count += 1
+                continue
+            if self._activity_last_query_start is None or row['query_start'] > self._activity_last_query_start:
+                self._activity_last_query_start = row['query_start']
+            yield row
+        if insufficient_privilege_count > 0:
+            self._log.warning(
+                "Insufficient privilege for %s/%s queries when collecting from %s.", self._config.pg_stat_activity_view
+            )
+            self._check.count(
+                "dd.postgres.statement_samples.error",
+                insufficient_privilege_count,
+                tags=self._tags + ["error:insufficient-privilege"],
+            )
+
+    def _get_db(self):
+        # while psycopg2 is threadsafe (meaning in theory we should be able to use the same connection as the parent
+        # check), the parent doesn't use autocommit and instead calls commit() and rollback() explicitly, meaning
+        # it can cause strange clashing issues if we're trying to use the same connection from another thread here.
+        # since the statement sampler runs continuously it's best we have our own connection here with autocommit
+        # enabled
+        if not self._db or self._db.closed:
+            self._db = self._check._new_connection()
+            self._db.set_session(autocommit=True)
+        if self._db.status != psycopg2.extensions.STATUS_READY:
+            # Some transaction went wrong and the connection is in an unhealthy state. Let's fix that
+            self._db.rollback()
+        return self._db
+
+    def _collection_loop(self):
+        try:
+            self._log.info("Starting statement sampler collection loop")
+            while True:
+                if time.time() - self._last_check_run > self._config.min_collection_interval * 2:
+                    self._log.info("Sampler collection loop stopping due to check inactivity")
+                    self._check.count("dd.postgres.statement_samples.collection_loop_inactive_stop", 1, tags=self._tags)
+                    break
+                self._collect_statement_samples()
+        except psycopg2.errors.DatabaseError as e:
+            self._log.warning(
+                "Statement sampler database error: %s", e, exc_info=self._log.getEffectiveLevel() == logging.DEBUG
+            )
```

Comment:
Why not just log as debug?

Reply:
I think since database errors will be relatively infrequent, it'll be useful for the user to (by default) see warnings in the log so they have context as to why it's failing (i.e. failed to connect). I think they shouldn't need to enable debug logs to see something that obvious. 

If debug is enabled then we'll log the full stack trace as well as if you're working on the code you may want to know exactly where the exception came from. But in the common case of something like failed connections, we don't need to pollute the log with the full stack trace.